app/src/main/java/io/appium/settings/recorder/RecorderThread.java

```diff
@@ -0,0 +1,466 @@
+import static io.appium.settings.recorder.RecorderConstant.NO_TIMESTAMP_SET;
+import static io.appium.settings.recorder.RecorderConstant.NO_TRACK_INDEX_SET;
+import static io.appium.settings.recorder.RecorderConstant.RECORDING_PRIORITY_DEFAULT;
+
+public class RecorderThread implements Runnable {
+
+    private static final String TAG = "RecorderThread";
+
+    private final MediaProjection mediaProjection;
+    private final String outputFilePath;
+    private final int videoWidth;
+    private final int videoHeight;
+    private final int recordingRotation;
+    private final int recordingPriority;
+    private final int recordingMaxDuration;
+
+    private boolean muxerStarted = false;
+    private boolean isStartTimestampInitialized = false;
+    private long startTimestampUs = System.nanoTime() / NANOSECOND_TO_MICROSECOND;
+    private long lastAudioTimestampUs = NO_TIMESTAMP_SET;
+
+    private int videoTrackIndex = NO_TRACK_INDEX_SET;
+    private int audioTrackIndex = NO_TRACK_INDEX_SET;
+
+    private volatile boolean stopped = false;
+    private volatile boolean audioStopped = false;
+    private volatile boolean hasAsyncError = false;
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private final VirtualDisplay.Callback displayCallback = new VirtualDisplay.Callback() {
+        @Override
+        public void onPaused() {
+            super.onPaused();
+            Log.v(TAG, "VirtualDisplay callback: Display streaming paused");
+        }
+
+        @Override
+        public void onStopped() {
+            super.onStopped();
+            if (!stopped) {
+                hasAsyncError = true;
+            }
+        }
+    };
+
+    public RecorderThread(MediaProjection mediaProjection, String outputFilePath,
+                          int videoWidth, int videoHeight, int recordingRotation,
+                          int recordingPriority, int recordingMaxDuration) {
+        this.mediaProjection = mediaProjection;
+        this.outputFilePath = outputFilePath;
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.recordingRotation = recordingRotation;
+        this.recordingPriority = recordingPriority;
+        this.recordingMaxDuration = recordingMaxDuration;
+    }
+
+    public void startRecording() {
+        stopped = false;
+        Thread recordingThread = new Thread(this);
+        recordingThread.start();
+    }
+
+    public void stopRecording() {
+        stopped = true;
+    }
+
+    public boolean isRecordingRunning() {
+        return !stopped;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
+    private MediaFormat initVideoEncoderFormat(String videoMime, int videoWidth,
+                                                int videoHeight, int videoBitrate) {
+        MediaFormat encoderFormat = MediaFormat.createVideoFormat(videoMime, videoWidth,
+                videoHeight);
+        encoderFormat.setInteger(MediaFormat.KEY_COLOR_FORMAT,
```

Comment:
If setInteger returns `this` then we could simply chain all calls:

```
return MediaFormat.createVideoFormat(videoMime, videoWidth, videoHeight)
  .setInteger()
  ....
```

Reply:
i would have love to do this but unfortunately `setInteger` is void type function